### DISCUSSION TITLE
Peer to peer example

### DISCUSSION THREAD

**Discussion by kennetpostigo** [webrtc-rs/webrtc#343] [Q&A] [answered]
Hi, i'm relatively new to webrtc. I wanted to ask if there is an example of what the rust side would look like for 2 clients to connect to the rust server in order to connect and transmit audio/video to each other in the browser?

Is there an example already for this?

**Comment by k0nserv** [comment, marked as answer]
Hey @kennetpostigo 

I think the [broadcast example](https://github.com/webrtc-rs/webrtc/tree/master/examples/examples/broadcast) is the best place to start for this. 

There's also a [WIP SFU example](https://github.com/webrtc-rs/examples/pull/15) from @sevensidedmarble and a more complete [SFU](https://github.com/billylindeman/switchboard) that @billylindeman has written.

**Reply by kennetpostigo** [reply]
@k0nserv what is an SFU?(Sorry mostly a webRTC noob, and trying to understand the moving parts) I've put together an overview of what I'm hoping to use the webrtc crate for:
<img width="1167" alt="image" src="https://user-images.githubusercontent.com/8888991/201781241-4ea29679-7fe4-46a5-963d-4e612e6ec1a8.png">

I think the overview is somewhat accurate, but if there are any corrections in how the webRTC works i'd love corrections! 

From what i've gathered I need:
1. Signaling Server to let 2 peers signal to each other and gather session description data.
2. Relay Server(from what I understand a fallback mechanism) to send media over TCP(?)
3. Stun Server to send data and media through to each connected client/peer.

**Reply by k0nserv** [reply]
Sorry about not being clearer. An [SFU](https://bloggeek.me/webrtcglossary/sfu/), Selective Forwarding Unit, is a server or mesh of servers that forward media tracks between peers selectively. It sounds like that's not what you want though. " I wanted to ask if there is an example of what the rust side would look like for 2 clients to connect to the rust server in order to connect and transmit audio/video to each other in the browser?", but it sounds like you want the peers to connect directly to each other?

You are mostly right about your points, except for #3. A [STUN](https://bloggeek.me/webrtcglossary/stun/) server is used for NAT hole punching i.e. to establish a network path between the peers, but network traffic then flows directly form peer to peer. Only when using a relay server([TURN](https://bloggeek.me/webrtcglossary/turn/)) does the traffic flow through another server.

An option is to establish a data channel between the peers in which case you can get away with using the signalling server only for the initial exchange, after which you can use the data channel for further communication.
